p2p/host/basic/addrs_reachability_tracker.go

```diff
@@ -0,0 +1,693 @@
+	r.mx.Unlock()
+	return reachable, unreachable
+}
+
+func (r *addrsReachabilityTracker) notify() {
+	select {
+	case r.reachabilityUpdateCh <- struct{}{}:
+	default:
+	}
+}
+
+func (r *addrsReachabilityTracker) updateTrackedAddrs(addrs []ma.Multiaddr) {
+	addrs = slices.DeleteFunc(addrs, func(a ma.Multiaddr) bool {
+		return !manet.IsPublicAddr(a)
+	})
+	if len(addrs) > maxTrackedAddrs {
+		log.Errorf("too many addresses (%d) for addrs reachability tracker; dropping %d", len(addrs), len(addrs)-maxTrackedAddrs)
+		addrs = addrs[:maxTrackedAddrs]
+	}
+	r.addrTracker.UpdateAddrs(addrs)
+}
+
+// reachabilityTask is a task to refresh reachability.
+// Waiting on the zero value blocks forever.
+type reachabilityTask struct {
+	Cancel context.CancelFunc
+	RespCh chan bool
+}
+
+func (r *addrsReachabilityTracker) refreshReachability() reachabilityTask {
+	if len(r.addrTracker.GetProbe()) == 0 {
+		return reachabilityTask{}
+	}
+	resCh := make(chan bool, 1)
+	ctx, cancel := context.WithTimeout(r.ctx, 5*time.Minute)
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		defer cancel()
+		backoff := runProbes(ctx, r.maxConcurrency, r.addrTracker, r.cli)
+		resCh <- backoff
+	}()
+	return reachabilityTask{Cancel: cancel, RespCh: resCh}
+}
+
+var errTooManyConsecutiveFailures = errors.New("too many consecutive failures")
+
+// errCountingClient counts errors from autonatv2Client and wraps the errors in response with a
+// errTooManyConsecutiveFailures in case of persistent failures from autonatv2 module.
+type errCountingClient struct {
+	autonatv2Client
+	MaxConsecutiveErrors int
+	mx                   sync.Mutex
+	consecutiveErrors    int
+}
+
+func (c *errCountingClient) GetReachability(ctx context.Context, reqs []autonatv2.Request) (autonatv2.Result, error) {
+	res, err := c.autonatv2Client.GetReachability(ctx, reqs)
+	c.mx.Lock()
+	defer c.mx.Unlock()
+	if err != nil && !errors.Is(err, context.Canceled) { // ignore canceled errors, they're not errors from autonatv2
+		c.consecutiveErrors++
+		if c.consecutiveErrors > c.MaxConsecutiveErrors {
+			err = fmt.Errorf("%w:%w", errTooManyConsecutiveFailures, err)
+		}
+		if errors.Is(err, autonatv2.ErrPrivateAddrs) {
+			log.Errorf("private IP addr in autonatv2 request: %s", err)
+		}
+	} else {
+		c.consecutiveErrors = 0
+	}
+	return res, err
+}
+
+type probeResponse struct {
+	Req []autonatv2.Request
+	Res autonatv2.Result
+	Err error
+}
+
+const maxConsecutiveErrors = 20
+
+// runProbes runs probes provided by addrsTracker with the given client. It returns true if the caller should
+// backoff before retrying probes. It stops probing when any of the following happens:
+// - there are no more probes to run
+// - context is completed
+// - there are too many consecutive failures from the client
+// - the client has no valid peers to probe
+func runProbes(ctx context.Context, concurrency int, addrsTracker *probeManager, client autonatv2Client) bool {
```

Comment:
This feels like it should be a method on addrsReachabilityTracker after
probeManager is consolidated to it.

Reply:
Okay, can we make this a method on `probeManager` at least?